zerver/views/__init__.py

```diff
@@ -1291,3 +1292,23 @@ def email_unsubscribe(request, type, token):
 
     return render_to_response('zerver/unsubscribe_link_error.html', {},
                               context_instance=RequestContext(request))
+
+# Custom realm emoji upload and deletion
+
+def list_emoji(request, user_profile):
+    return json_success({'emoji': user_profile.realm.get_emoji()})
+
+def upload_emoji(request, user_profile):
+    emoji_name = request.POST.get('name', None)
+    emoji_url = request.POST.get('url', None)
+    try:
+        do_add_realm_emoji(user_profile.realm, emoji_name, emoji_url)
+    except ValidationError as e:
+        if '__all__' in e.message_dict:
+            return json_error(e.message_dict['__all__'])
```

Comment:
Sorry, missed that one. This happens when you try to add an emoji with a duplicate name, the error looks like: `Realm emoji with this Realm and Name already exists.`